man/man8/mpp88.8
.TH MPP88 8
.\" @(#) $Header: /xtel/pp/pp-beta/man/man8/RCS/mpp88.8,v 6.0 1991/12/18 20:44:03 jpo Rel $
.\"
.\" $Log: mpp88.8,v $
.\" Revision 6.0  1991/12/18  20:44:03  jpo
.\" Release 6.0
.\"
.\"
.\"
.SH NAME
mpp88 \- message pretty printer
.SH
.in +.5i
.ti -.5i
mpp88 \%[-ps] \%[-pl] \%[-p1] \%[-p2] \%[-ipn] \%[-nounpack] \%[-p2hdr]
.in -.5i
.SH DESCRIPTION
.I Mpp88
is a diagnostic tool for examining X.4001988) message structures. It
handles various types of structures either in raw asn.1 or in the
presentation list abstraction of ISODE.
.I Mpp88
responds to the following flags:
.TP
\-ps
indicates that the input is from a presentation stream abstraction
(e.g raw ASN.1)
.TP
\-pl
indicates that the input is from a presentation list abstraction.
.TP
\-p1
Interpret the data as a P1 envelope with P2 encapsualted within.
.TP
\-p2
Interpret the data as a P2 header and body parts.
.TP
\-ipn
Interpret the data as a IPN.
.TP
\-nounpack
Do not recursively unpack sub structures encoded as OCTET STRINGs.
.TP
\-p2hdr
Interpret the data as a p2 header sequence.
.SH FILES
none
.SH "SEE ALSO"
mpp84(8)
.br
\fIThe PP Manual: Volume 1 \- Installation and Operation\fP



CHANGES









             Changes from PP version 5.2 to 6.0





_1.  _O_v_e_r_v_i_e_w

This  short  document  is  a   description   of   the   more
user/administrator  visible changes that have taken place in
PP between the last interim version (5.2) and the beta  ver-
sion  (6.0). This document is not exhaustive or complete but
is a guide to what to look out for. In general the following
may be assumed.

o+    The Queue files are compatible  between  5.2  and  6.0.
     Note  however that the Queue structure has changed. Old
     channels will have to be recompiled  to  recognise  the
     new format, but new channels will read old queue files.

o+    The qmgr protocol is compatible between 5.2 and 6.0

o+    The syntax of some of  the  table  files  has  changed.
     Where appropriate conversion scripts are provided.

o+    Some procedure calls have been  changed  or  have  dif-
     ferent parameters.

o+    the libdsap.a library is now required for compilation.

In summary, you should be careful when upgrading to the  6.0
release.   It  will  not  work  by  just  replacing  the 5.2
binaries with 6.0. For definite the domain table  will  HAVE
to be changed.

_2.  _D_e_t_a_i_l_e_d _c_h_a_n_g_e_s

This list is in no particular order of preference or  impor-
tance.

Domain Table
     The domain table syntax has been  changed  to  be  more
     explicit  and  has  been  brought into line with the OR
     table syntax. We feel this is clearer and more logical.
     Tools  that  process  the domain table have been taught
     the   new   format   and   there   are    scripts    in
     Tools/tables/compat  to convert between the old and new
     formats.  There is NO backwards compatibility here. Old
     format tables will just plain not work.

Aliases Table



                     December 18, 1991





                           - 2 -


     The aliases table syntax has changed to a  main  choice
     of  synonym or alias with further qualifiers specifying
     the address type of the synonym or alias.  The  changes
     are

          x:alias y         x:alias y
          x:synonym y       x:synonym y
          x:822-synonym y   x:synonym y 822
          x:x400-synonym y  x:synonym y x400
          x:822 y           x:alias y 822
          x:x400 y          x:alias y x400
     where any white space in  y  must  be  quoted  using  a
     backslash  There  is also a "external" qualifier.  This
     disables further lookups of synonyms  and  aliases  and
     can  be  used  for  a  main  site mail hub mapping from
     site_wide_name@localsite                             to
     local_machine_name@localmachine  and  vice  versa.  The
     use of external stops the looping alias  problem  which
     would otherwise occur.

Return of Contents
     Return of contents is now supported.  This  is  a  side
     effect  and  part  of  the  work on routing of delivery
     reports.

Use of pepsy
     Use of the PEPSY ASN.1 compiler rather  than  PEPY/POSY
     is  now  used  in  many  places. This makes for smaller
     binaries and much quicker compilation.

ISODE 7.0
     PP 6.0 is aligned to ISODE 7.0. It will not work  under
     ISODE 6.0.  If you are using the 6.0 release you should
     be using ISODE 7.0. It should however work under  ISODE
     6.8.

Tailor file
     There have been many changes to the  tailor  file.  The
     introduction of several new variables and the splitting
     up of header and body part types. These are

          dap_user     Directory username for DS lookups
          dap_passwd   Directory Passwd for same
          fsync        disable use of fsync(2)
          headertype   list of headers c.f. bodypart
          queuestruct  use hierarchical queue structure
          submit_addr  call submit over network
          diskuse      limit the disk space consumed by PP

     Channel tailoring also has a number of new variables

          drchan       replaces old outchan
          hdrin        headers inbound
          hdrout       headers outbound



                     December 18, 1991





                           - 3 -


          ininfo       inbound channel info
          inadr        inbound channel addr type
          insubadr     inbound channel sub address type
          intable      inbound channel table
          key          give alternative names to channel
          mtatable     list of known mtas for channel binding
          outinfo      outbound channel info
          outsubadr    outbound channel sub address
          outtable     outbound channel table
          solo-proc    channel requires address one at a time
          bad-sender-policyset the policy for unroutable senders
          check        set the checking mode of the channel

Channel pairing
     Channels may now be explicitly  paired  together.  This
     allows one definition to specify both an inbound and an
     outbound channel.  It also allows several logical chan-
     nels  for  one  inbound protocol by selecting a channel
     per MTA.

Timezones
     Many more lettered timezones are recognised. This  list
     is ambiguous at best and numeric forms should really be
     used except at user interfaces.

Delivery Reports
     Delivery reports are now routed back to their  origina-
     tor.  In 5.2 they were returned directly down the chan-
     nel they arrived on. This allows for one way  links  in
     general  makes  things  easier. It also allows protocol
     conversion so that, for instance, a delivery report  on
     an  X.400 message may go back via an RFC-822 if that is
     the only option.  A side effect of this is the  channel
     variable  outchan  has been renamed drchan and has dif-
     ferent semantics attached to it (see the manual).

Header parts
     Separation of headers and bodyparts for  each  channel.
     This  is  mainly a syntactic change but will allow more

Tools/mpp88/mpp88.py
-- mpp88.py: - test out PEPY


-- @(#) $Header: /xtel/pp/pp-beta/Tools/mpp88/RCS/mpp88.py,v 6.0 1991/12/18 20:31:43 jpo Rel $
--
-- $Log: mpp88.py,v $
-- Revision 6.0  1991/12/18  20:31:43  jpo
-- Release 6.0
--
--
--


MPP DEFINITIONS   ::=

%{
#ifndef lint
static char *rcsid = "$Header: /xtel/pp/pp-beta/Tools/mpp88/RCS/mpp88.py,v 6.0 1991/12/18 20:31:43 jpo Rel $";
#endif  lint

#include <stdio.h>
#include "Trans-types.h"
#include "IOB-types.h"


#define ps_advise(ps, f) \
        advise (NULLCP, "%s: %s", (f), ps_error ((ps) -> ps_errno))

/*    DATA */
int		unpack = 1;
static char	*myname = "mpp88";
static enum	{ ps2mpp, pl2mpp } mode = ps2mpp;
static enum	format { p1, p2, p2hdr, ipn} topfmt = p2;
void		adios ();


/*    MAIN */

/* ARGSUSED */

main (argc, argv, envp)
int     argc;
char  **argv,
      **envp;
{
    register int    status = 0;
    register char  *cp;
    register FILE  *fp;

    if (myname = rindex (argv[0], '/'))
        myname++;
    if (myname == NULL || *myname == NULL)
        myname = argv[0];

    for (argc--, argv++; cp = *argv; argc--, argv++)
        if (*cp == '-') {
            if (strcmp (cp + 1, "ps") == 0) {
                mode = ps2mpp;
                continue;
            }
            if (strcmp (cp + 1, "pl") == 0) {
                mode = pl2mpp;
                continue;
            }
            if (strcmp (cp + 1, "p1") == 0) {
                topfmt = p1;
                continue;
            }
            if (strcmp (cp + 1, "p2") == 0) {
                topfmt = p2;
                continue;
            }
	    if (strcmp (cp + 1, "nounpack") == 0) {
		unpack = 0;
		continue;
	    }
	    if (strcmp (cp + 1, "p2hdr") == 0) {
		topfmt = p2hdr;
		continue;
	    }
	    if (strcmp (cp + 1, "ipn") == 0) {
		topfmt = ipn;
		continue;
	    }
            adios (NULLCP, "usage: %s [ -ps | -pe ] [-p1 | -p2 | -p2hdr | -ipn ] [ files... ]",
                    myname);
        }
        else
            break;

    if (argc == 0)
        status = process ("(stdin)", stdin);
    else
        while (cp = *argv++) {
            if ((fp = fopen (cp, "r")) == NULL) {
                advise (cp, "unable to read");
                status++;
                continue;
            }
            status += process (cp, fp);
            (void) fclose (fp);
        }

    exit (status);              /* NOTREACHED */
}

/*  */

static int  process (file, fp)
register char *file;
register FILE *fp;
{
    enum format curfmt = topfmt;
    register PE     pe;
    register PS     ps;

    if ((ps = ps_alloc (std_open)) == NULLPS) {
        ps_advise (ps, "ps_alloc");
        return 1;
    }
    if (std_setup (ps, fp) == NOTOK) {
        advise (NULLCP, "%s: std_setup loses", file);
        return 1;
    }

    for (;;) {
        switch (mode) {
            case ps2mpp: 
                if ((pe = ps2pe (ps)) == NULLPE)
                    if (ps -> ps_errno) {
                        ps_advise (ps, "ps2pe");
                you_lose: ;
			if (pe != NULLPE)
				pe_free (pe);
			ps_free (ps);
                        return 1;
                    }
                    else {
                done:   ;
                        ps_free (ps);
                        return 0;
                    }
                break;

            case pl2mpp: 
                if ((pe = pl2pe (ps)) == NULLPE)
                    if (ps -> ps_errno) {
                        ps_advise (ps, "pl2pe");
                        goto you_lose;
                    }
                    else
                        goto done;
                break;
        }

        switch (curfmt) {
            case p1:
            default:
		(void) print_Trans_MtsAPDU (pe, 1, 0, NULLCP, NULL);
                break;

            case p2:
                (void) print_IOB_InformationObject (pe, 1, NULLIP, NULLVP, NULLCP);
                break;
	    case p2hdr:
		(void) print_IOB_Heading (pe, 1, NULLIP, NULLVP, NULLCP);
		break;
	    case ipn:
		(void) print_IOB_IPN (pe, 1, NULLIP, NULLVP, NULLCP);
		break;
        }

        pe_free (pe);
    }
}

/*  */

%}

BEGIN

END

%{
/*    DEBUG */

#ifdef  DEBUG
testdebug (pe, s)
register PE pe;
register char *s;
{
    static int  debug = OK;
    char   *cp;
    register PS     ps;

    switch (debug) {
        case NOTOK: 
            break;

        case OK: 
            if ((debug = (cp = getenv ("PEPYDEBUG")) && *cp ? atoi (cp)
                        : NOTOK) == NOTOK)
                break;
            fprintf (stderr, "%s made with %s\n", myname, pepyid);

        default: 
            fprintf (stderr, "%s\n", s);

            if ((ps = ps_alloc (std_open)) == NULLPS)
                return;
            if (std_setup (ps, stderr) != NOTOK)
                (void) pe2pl (ps, pe);
            fprintf (stderr, "--------\n");
            ps_free (ps);
            break;
    }
}
#endif  DEBUG

/*    ERRORS */

/* VARARGS2 */

void    adios (what, fmt, a, b, c, d, e, f, g, h, i, j)
char   *what,
       *fmt,
       *a,
       *b,
       *c,
       *d,
       *e,
       *f,
       *g,
       *h,
       *i,
       *j;
{
    advise (what, fmt, a, b, c, d, e, f, g, h, i, j);
    _exit (1);
}

/*  */

/* VARARGS2 */

void    advise (what, fmt, a, b, c, d, e, f, g, h, i, j)
char   *what,
       *fmt,
       *a,
       *b,
       *c,
       *d,
       *e,
       *f,
       *g,
       *h,
       *i,
       *j;
{
    (void) fflush (stdout);

    fprintf (stderr, "%s: ", myname);
    fprintf (stderr, fmt, a, b, c, d, e, f, g, h, i, j);
    if (what)
        (void) fputc (' ', stderr), perror (what);
    else
        (void) fputc ('\n', stderr);
    (void) fflush (stderr);
}

%}
